lib/krikri/harvesters/oai_harvester.rb

```diff
@@ -51,38 +76,107 @@ def count
     #
     #     records.take(1000)
     #
+    # @param opts [Hash] opts to pass to OAI::Client
+    # @see #expected_opts
     def records(opts = {})
-      opts = opts.merge(@opts)
-      client.list_records(opts).full.lazy.flat_map do |rec|
-        @record_class.build(mint_id(rec.header.identifier),
-                            record_xml(rec))
-
+      opts = @opts.merge(opts)
+      request_with_sets(opts) do |set_opts|
+        client.list_records(set_opts).full.lazy.flat_map do |rec|
+          @record_class.build(mint_id(rec.header.identifier),
+                              record_xml(rec))
+        end
       end
+      # sets = Array(opts.delete(:set) { [nil] })
+
+      # set_enums = sets.map do |set|
+      #   set_opts = opts.dup
+      #   set_opts[:set] = set unless set.nil?
+      #   client.list_records(set_opts).full.lazy.flat_map do |rec|
+      #     @record_class.build(mint_id(rec.header.identifier),
+      #                         record_xml(rec))
+      #   end
+      # end
+      # concat_enum(*set_enums).lazy
     end
 
-    # TODO: normalize records; there will be differences in XML
-    # for different requests
+    ##
+    # Gets a single record with the given identifier from the OAI endpoint
+    #
+    # @param identifier [#to_s] the identifier of the record to get
+    # @param opts [Hash] options to pass to the OAI client
     def get_record(identifier, opts = {})
       opts[:identifier] = identifier
-      opts = opts.merge(@opts)
+      opts = @opts.merge(opts)
       @record_class.build(mint_id(identifier),
                           record_xml(client.get_record(opts).record))
     end
 
     ##
+    # Lists the sets available from the OAI endpoint. Accepts a block to
+    # pass to `#map` on the resulting array.
+    #
+    # @example:
+    #
+    #   sets(&:spec)
+    #
+    # @param opts [Hash] options to pass to the OAI client
+    # @return [Array<OAI::Set>] an array of sets.
+    #
+    # @see OAI::Set
+    def sets(opts = {}, &block)
+      arry = client.list_sets.full.to_a
+      return arry unless block_given?
+      arry.map(&block)
+    end
+
+    ##
+    # @return [Hash] A hash documenting the allowable options to pass to
+    #   initializers.
+    #
     # @see Krikri::Harvester::expected_opts
     def self.expected_opts
       {
         key: :oai,
         opts: {
           set: {type: :string, required: false, multiple_ok: true},
-          metadata_prefix: {type: :string, required: true}
+          skip_set: {type: :string, required: false, multiple_ok: true},
+          metadata_prefix: {type: :string, required: false},
+          from: {type: :string, required: false},
+          until: {type: :string, required: false}
         }
       }
     end
 
+    ##
+    # Concatinates two enumerators
+    # TODO: find a better home for this. Reopen Enumerable?
+    def concat_enum(*enums)
+      Enumerator.new do |yielder|
```

Comment:
Would it be possible to do something like `enums.flat_map { |e| e }`, or am I missing something about the seeming simplicity of that?